examples/virtual-dom-wip/platform/Html/Internal/Client.roc

```diff
@@ -0,0 +1,538 @@
+                    rendered: renderedWithHandler,
+                    patches: List.append walkState.patches patch,
+                }
+
+createChildNode :
+    { rendered : RenderedTree state, patches : List Patch, ids : List NodeId },
+    Html state
+    -> { rendered : RenderedTree state, patches : List Patch, ids : List NodeId }
+createChildNode = \{ rendered, patches, ids }, childHtml ->
+    { rendered: renderedChild, patches: childPatches, id } =
+        createNode { rendered, patches } childHtml
+
+    {
+        rendered: renderedChild,
+        patches: childPatches,
+        ids: List.append ids id,
+    }
+
+# insert a node into the nodes list, assigning it a NodeId
+insertNode : RenderedTree state, RenderedNode -> { rendered : RenderedTree state, id : NodeId }
+insertNode = \rendered, node ->
+    when List.last rendered.deletedNodeCache is
+        Ok id ->
+            newRendered =
+                { rendered &
+                    nodes: List.set rendered.nodes id (Ok node),
+                    deletedNodeCache: List.dropLast rendered.deletedNodeCache,
+                }
+
+            { rendered: newRendered, id }
+
+        Err _ ->
+            newRendered =
+                { rendered &
+                    nodes: List.append rendered.nodes (Ok node),
+                }
+
+            { rendered: newRendered, id: List.len rendered.nodes }
+
+# Predict what NodeId will be assigned next, without actually assigning it
+nextNodeId : RenderedTree state -> NodeId
+nextNodeId = \rendered ->
+    when List.last rendered.deletedNodeCache is
+        Ok id -> id
+        Err _ -> List.len rendered.nodes
+
+# -------------------------------
+#   TESTS
+# -------------------------------
+# indexNodes
+expect
+    html : Html {}
+    html =
+        Element "a" 43 [HtmlAttr "href" "https://www.roc-lang.org/"] [Text "Roc"]
+
+    actual : { nodes : List RenderedNode, siblingIds : List Nat }
+    actual =
+        indexNodes { nodes: [], siblingIds: [] } html
+
+    expected : { nodes : List RenderedNode, siblingIds : List Nat }
+    expected = {
+        nodes: [
+            RenderedText "Roc",
+            RenderedElement "a" [RenderedHtmlAttr "href" "https://www.roc-lang.org/"] [0],
+        ],
+        siblingIds: [1],
+    }
+
+    actual == expected
+
+# initClientAppHelp
+expect
+    State : { answer : Nat }
+
+    init = \result ->
+        when result is
+            Ok state -> state
+            Err _ -> { answer: 0 }
+
+    onClickHandler : Handler State
+    onClickHandler =
+        Normal \state, _ -> Action.update { answer: state.answer + 1 }
+
+    render : State -> Html State
+    render = \state ->
+        num = Num.toStr state.answer
+
+        onClickAttr : Attribute State
+        onClickAttr =
+            EventListener "click" [] onClickHandler
+
+        Element "body" 0 [] [
+            Element "h1" 0 [] [Text "The app"],
+            Element "div" 0 [onClickAttr] [Text "The answer is \(num)"],
+        ]
```

Comment:
Is it fine that the sizes here are 0? Should they be material?

Reply:
They only matter on the server side and this is a client side test so I didn't bother.

Reply:
I could probably make that clearer though